multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,327 @@
+			return r.onPodDelete(ctx, req.NamespacedName.String(), true)
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	normalizedLabel := getNormalizedLabel(ns.Labels, pod.Labels, ns.Name)
+	return r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), normalizedLabel)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(predicate.LabelChangedPredicate{}).
+		Watches(&source.Kind{Type: &v1.Namespace{}}, handler.EnqueueRequestsFromMapFunc(r.namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// onPodDelete removes the Pod and label identity mapping from the cache, and deletes label identity
+// ResourceExport if necessary (the Pod deletion event causes a label identity no longer present in
+// the cluster).
+// Note that if releaseLabelMutex is set to false, r.labelMutex will continue to be held after the
+// function returns, which is useful for eliminating additional context switching when handling Pod
+// update events, since after deleting stale label the labelMutex will immediately need to be held
+// again for checking label add.
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string, releaseLabelMutex bool) (ctrl.Result, error) {
+	r.labelMutex.Lock()
+	if originalLabel, isCached := r.podLabelCache[podNamespacedName]; isCached {
+		if requeue := r.requeueIfLabelResExportUpdateInProgress(originalLabel); requeue {
+			r.labelMutex.Unlock()
+			return ctrl.Result{Requeue: true}, nil
+		}
+		// Check if the original label is stale.
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
+			klog.V(2).InfoS("Label no longer exists in the cluster, deleting corresponding ResourceExport", "label", originalLabel)
+			if err := r.deleteLabelIdentityResExport(ctx, originalLabel); err != nil {
+				r.labelMutex.Unlock()
```

Comment:
We need to unlock before create/delete Exports.

Reply:
If you check `deleteLabelIdentityResExport()`, the labelMutex is released before making the API call, and acquired again immediately after. But I also realized that in case of API failure there's no need for the lock to be acquired again. Will change so that it is clearer.